src/compile/ast.rs

```diff
@@ -913,66 +913,75 @@ impl Type {
+                Value::from(0), // set size to 0
+            ]),
             Type::Array(t) => {
-                let (def, safe) = t.default_value();
-                (
-                    Value::new_tuple(vec![
-                        Value::Int(Uint256::one()),
-                        Value::Int(Uint256::one()),
-                        Value::new_tuple(vec![def]),
-                    ]),
-                    safe,
-                )
+                let fixed = Type::FixedArray(t.clone(), 1).default_value(type_tree);
+                Value::new_tuple(vec![
+                    Value::from(1), // size
+                    Value::from(1), // topstep
+                    fixed,          // array.mini builtin_arrayNew() unsafe casts this
+                ])
             }
-            Type::FixedArray(t, sz) => {
-                let (default_val, safe) = t.default_value();
-                let mut val = Value::new_tuple(vec![default_val; 8]);
-                let mut chunk_size = 1;
-                while chunk_size * TUPLE_SIZE < *sz {
-                    val = Value::new_tuple(vec![val; 8]);
-                    chunk_size *= 8;
+            Type::FixedArray(t, size) => {
+                // emulate array.mini builtin_arrayNew()
+                fn emulated_builtin(size: usize, mut base: Value) -> Value {
+                    let mut chunk = 1;
+                    while (8 * chunk < size) {
+                        chunk = 8 * chunk;
+                        base = Value::new_tuple(vec![base; 8]);
+                    }
+                    Value::new_tuple(vec![
+                        Value::from(size),  // size
+                        Value::from(chunk), // topstep
+                        Value::new_tuple(vec![base; 8]),
+                    ])
                 }
-                (val, safe)
+                emulated_builtin(*size, t.default_value(type_tree))
             }
-            Type::Struct(fields) => {
-                let mut vals = Vec::new();
-                let mut is_safe = true;
-                for field in fields {
-                    let (val, safe) = field.tipe.default_value();
-                    vals.push(val);
-                    is_safe = is_safe && safe;
-                }
-                (value_from_field_list(vals), is_safe)
+            Type::Nominal(..) => {
+                let tipe = self.rep(type_tree).unwrap_or(Type::Any);
+                tipe.default_value(type_tree)
             }
-            Type::Map(_, _) | Type::Func(_, _, _) | Type::Nominal(_, _, _) => {
-                (Value::none(), false)
+            Type::Generic(_) | Type::GenericSlot(_) => {
+                // This can happen with builtins, but we plan on refactoring all builtins out.
+                // In the future we'll make this panic since you should always know the specified value
+                // for whatever generic type you're getting the default for.
+                Value::none()
```

Comment:
Unless there is some rationale I am missing, it might be better to keep a safety marker in the return type if unsafe defaults are still possible. We can warn if something like this is present on a type if this marker is there.

Reply:
The rationale is that, although this function used to return a safety marker, it was always underscored and discarded. `NewArray`'s codegen was the only caller and, since the type tree is lost at this point, we've been filling it with empty tuples. Now we do have `link()` as the additional caller, but, since globals always have a specialized version of their generic types, you can't have any `Generic` or `GenericSlot` sub-types. So the only scenario where this would cause a bug is one where we already have prior to this PR.

You're hitting on something good here, though, which is that we should move to not using builtins since their presence edge-cases the safety guarantees of constructs like generics. Once we remove `newarray` in specific, we can make this function panic on these two types.